### PR TITLE
 Rework Step information to use logging infrastructure #1106 

### PR DESCRIPTION
**Pull request by Emantor**
**Description**
Hopefully the final iteration of the logging rework.

**Checklist**

- [x] PR has been tested

Closes #1106

### CODE REVIEW

File: labgrid/logging.py [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+
+logging.CONSOLE = logging.INFO - 5
+assert(logging.CONSOLE > logging.DEBUG)
+logging.addLevelName(logging.CONSOLE, "CONSOLE")
+
+# Use composition instead of inheritance
+class StepFormatter:
+    def __init__(
+        self, *args, indent=True, color=None, parent=None, **kwargs
+    ):
```
Comment (Bastian-Krause) [inline labgrid/logging.py:28]:
```suggestion
    def __init__(self, *args, indent=True, color=None, parent=None, **kwargs):
```

File: labgrid/logging.py [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+        if parent:
+            self.formatter = parent
+        else:
+            self.formatter = logging.Formatter(*args, **kwargs)
```
Comment (Bastian-Krause) [inline labgrid/logging.py:32]:
```suggestion
        self.formatter = parent or logging.Formatter(*args, **kwargs)
```

File: labgrid/logging.py [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+        self.indent = indent
+        self.indent_level = 0
+        self.bufs = dict()
+        self.re_vt100 = re.compile(r"(\x1b\[|\x9b)[^@-_a-z]*[@-_a-z]|\x1b[@-_a-z]")
```
Comment (Bastian-Krause) [inline labgrid/logging.py:36]: The same regex is used in the BareboxDriver, ShellDriver and UBootDriver. Maybe store it somewhere in util and use it from all of these drivers, including logging?

File: labgrid/logging.py [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+                record.msg = (" " * self.indent_level) + record.msg
+
+                self.indent_level = getattr(
+                    record, "next_indent_level", self.indent_level
+                )
+
+            if hasattr(record, "step"):
+                record.pathname = record.step.sourceinfo[0]
+                record.filename = record.step.sourceinfo[1]
+                record.lineno = record.step.sourceinfo[2]
```
Comment (Bastian-Krause) [inline labgrid/logging.py:55]: Maybe..
```suggestion
                record.pathname, record.filename,  record.lineno = record.step.sourceinfo
```

File: labgrid/logging.py [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+    )
+    loggers = attr.ib(
+        default=attr.Factory(dict), validator=attr.validators.instance_of(dict)
+    )
+    lastevent = attr.ib(
+        default=None,
+        validator=attr.validators.optional(attr.validators.instance_of(StepEvent)),
+    )
+    re_vt100 = attr.ib(
+        default=re.compile(r"(\x1b\[|\x9b)[^@-_a-z]*[@-_a-z]|\x1b[@-_a-z]")
```
Comment (Bastian-Krause) [inline labgrid/logging.py:75]: See above.

File: labgrid/logging.py [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+
+    @classmethod
+    def format_result(cls, result):
+        if result is None:
+            return ""
+
+        if len(str(result)) < cls._length_limit or cls._length_limit is None:
+            return "result={} ".format(result)
+        else:
+            return f"result={repr(result):.{cls._length_limit}}… "
```
Comment (Bastian-Krause) [inline labgrid/logging.py:215]:
```suggestion

        return f"result={repr(result):.{cls._length_limit}}… "
```

File: labgrid/pytestplugin/hooks.py [hunk 7]
```diff
@@ -1,30 +1,66 @@
...
 from ..consoleloggingreporter import ConsoleLoggingReporter
-from .reporter import StepReporter, ColoredStepReporter
 from ..util.helper import processwrapper
+from ..logging import StepFormatter, StepLogger
+
+@pytest.hookimpl(tryfirst=True)
+def pytest_cmdline_main(config):
+    def set_cli_log_level(l):
+        nonlocal config
+        current_level = getattr(config.option, "log_cli_level", None)
```
Comment (Bastian-Krause) [inline labgrid/pytestplugin/hooks.py:15]:
```suggestion
        current_level = getattr(config.option, "log_cli_level", logging.NOTSET)
```

File: labgrid/pytestplugin/hooks.py [hunk 8]
```diff
@@ -1,30 +1,66 @@
...
+            current_level = current_level.upper()
+            try:
+                current_level = int(getattr(logging, current_level, None))
+            except ValueError:
+                current_level = logging.WARNING
+
+        if current_level is None or \
+                current_level == logging.NOTSET or \
+                current_level > l:
+            config.option.log_cli_level = str(l)
```
Comment (Bastian-Krause) [inline labgrid/pytestplugin/hooks.py:27]:
```suggestion
            try:
                current_level = int(logging.getLevelName(current_level))
            except ValueError:
                current_level = logging.NOTSET

        if current_level == logging.NOTSET or current_level > l:
            config.option.log_cli_level = str(l)
```
As far as I understand the [docs](https://docs.pytest.org/en/7.1.x/reference/reference.html#confval-log_cli_level), we don't need to (and shouldn't) handle lower case log level names.

File: tests/test_steplogger.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
+import logging
```
Comment (Bastian-Krause) [inline tests/test_steplogger.py:1]: Unused.

File: tests/test_steplogger.py [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+import os
```
Comment (Bastian-Krause) [inline tests/test_steplogger.py:2]: Unused.

**Review by Bastian-Krause — CHANGES_REQUESTED** [review]
The changes in the Step class are weird:
- first, `Step.__str__` is dropped, Step.get_title is added
- then, `Step.__str__` is readded
- finally, `Step.get_title` and `Step.__str__` are dropped again

`labgrid/pytestplugin/reporter.py` is modified, then removed completely. I guess these commits can be squashed?

We should add a note that `--lg-colored-steps` has currently no effect and what's the plan for the future.

File: labgrid/driver/smallubootdriver.py [hunk 11]
```diff
@@ -6,6 +6,7 @@
 from .common import Driver
 
 from .ubootdriver import UBootDriver
+from .util import re_vt100
```
Comment (SmithChart) [inline labgrid/driver/smallubootdriver.py:9]: This import must be ``from ..util`` as with the ``ubootdriver.py``.

**Review by Bastian-Krause — APPROVED** [review]

**Pull request merged**